document-store/src/main/java/org/hypertrace/core/documentstore/postgres/PostgresCollection.java

```diff
@@ -0,0 +1,347 @@
+  private String columnType;
+  
+  public PostgresCollection(Connection client, String collectionName, String columnType) {
+    this.client = client;
+    this.collectionName = collectionName;
+    this.columnType = columnType;
+  }
+  
+  @Override
+  public boolean upsert(Key key, Document document) {
+    String jsonString = document.toJson();
+    String upsertSQL = String.format("INSERT INTO %s (%s,%s) VALUES( ?, ? :: %s) ON CONFLICT(%s) DO UPDATE SET %s = ?::%s ",
+      collectionName, ID_KEY, DOCUMENT_KEY, columnType, ID_KEY, DOCUMENT_KEY, columnType);
+    try {
+      
+      PreparedStatement preparedStatement = client.prepareStatement(upsertSQL, Statement.RETURN_GENERATED_KEYS);
+      preparedStatement.setString(1, key.toString());
+      preparedStatement.setString(2, jsonString);
+      preparedStatement.setString(3, jsonString);
+      ResultSet resultSet = preparedStatement.executeQuery();
+      
+      if (LOGGER.isDebugEnabled()) {
+        LOGGER.debug("Write result: " + resultSet.toString());
+      }
+      
+      return resultSet.rowUpdated();
+    } catch (SQLException e) {
+      LOGGER.error("SQLException inserting document. key: {} content:{}", key, document, e);
+    }
+    return false;
+  }
+  
+  @Override
+  public boolean updateSubDoc(Key key, String subDocPath, Document subDocument) {
+    return false;
+  }
+  
+  @Override
+  public Iterator<Document> search(Query query) {
+    String filters = null;
+    String space = " ";
+    StringBuilder searchSQLBuilder = new StringBuilder("SELECT * FROM")
+      .append(space).append(collectionName);
+    
+    // If there is a filter in the query, parse it fully.
+    if (query.getFilter() != null) {
+      filters = parseQuery(query.getFilter());
+    }
+    
+    LOGGER.debug(
+      "Sending query to PostgresSQL: {} : {}",
+      collectionName,
+      filters);
+    
+    if (filters != null) {
+      searchSQLBuilder
+        .append(" WHERE ").append(filters);
+    }
+    
+    if (!query.getOrderBys().isEmpty()) {
+      String orderBySQL = parseOrderByQuery(query.getOrderBys());
+      searchSQLBuilder.append(" ORDER BY ").append(orderBySQL);
+    }
+    
+    Integer limit = query.getLimit();
+    if (limit != null && limit >= 0) {
+      searchSQLBuilder.append(" LIMIT ").append(limit);
+    }
+    
+    Integer offset = query.getOffset();
+    if (offset != null && offset >= 0) {
+      searchSQLBuilder.append(" OFFSET ").append(offset);
+    }
+    
+    try {
+      PreparedStatement preparedStatement = client.prepareStatement(searchSQLBuilder.toString());
+      ResultSet resultSet = preparedStatement.executeQuery();
+      return new PostgresResultIterator(resultSet);
+    } catch (SQLException e) {
+      LOGGER.error("SQLException querying documents. query: {}", query, e);
+    }
+    
+    return null;
+  }
+  
+  @VisibleForTesting
+  /**
+   *
+   */
+  String parseQuery(Filter filter) {
+    if (filter.isComposite()) {
```

Comment:
I think this can break down into smaller static methods to improve readability something like this::parseQueryForCompositeFilter and this::parseQueryForNonCompositeFilter